packages/@react-spectrum/tag/src/TagGroup.tsx

```diff
@@ -10,50 +10,88 @@
  */
 
-import {classNames, useStyleProps} from '@react-spectrum/utils';
-import React, {useContext} from 'react';
-import {Removable} from '@react-types/shared';
+import {classNames, useDOMRef, useStyleProps} from '@react-spectrum/utils';
+import {DOMRef} from '@react-types/shared';
+import {GridCollection, useGridState} from '@react-stately/grid';
+import {GridKeyboardDelegate, useGrid} from '@react-aria/grid';
+import {mergeProps} from '@react-aria/utils';
+import React, {ReactElement, useMemo} from 'react';
 import {SpectrumTagGroupProps} from '@react-types/tag';
 import styles from '@adobe/spectrum-css-temp/components/tags/vars.css';
+import {Tag} from './Tag';
+import {useListState} from '@react-stately/list';
+import {useLocale} from '@react-aria/i18n';
 import {useProviderProps} from '@react-spectrum/provider';
 import {useTagGroup} from '@react-aria/tag';
 
-interface TagGroupContextValue extends Removable<any, void> {
-  isDisabled?: boolean,
-  isFocused?: boolean,
-  isRequired?: boolean,
-  isReadOnly?: boolean,
-  validationState?: 'valid' | 'invalid',
-  role?: 'gridcell'
-}
-
-const TagGroupContext = React.createContext<TagGroupContextValue | {}>({});
-
-export function useTagGroupProvider(): TagGroupContextValue {
-  return useContext(TagGroupContext);
-}
 
-export const TagGroup = ((props: SpectrumTagGroupProps) => {
+function TagGroup<T extends object>(props: SpectrumTagGroupProps<T>, ref: DOMRef<HTMLDivElement>) {
   props = useProviderProps(props);
-
   let {
-    isReadOnly,
     isDisabled,
+    isRemovable,
     onRemove,
-    validationState,
-    children,
     ...otherProps
   } = props;
+
+  let domRef = useDOMRef(ref);
   let {styleProps} = useStyleProps(otherProps);
+  let {direction} = useLocale();
   const {tagGroupProps} = useTagGroup(props);
+  let {collection} = useListState(props);
+  let gridCollection = useMemo(() => new GridCollection({
+    columnCount: isRemovable ? 2 : 1,
+    items: [...collection].map(item => {
+      let childNodes = [{
+        ...item,
+        index: 0,
+        type: 'cell'
+      }];
 
-  function removeAll(tags) {
-    onRemove([tags]);
-  }
+      // add column of clear buttons if removable
+      if (isRemovable) {
+        childNodes.push({
+          key: `remove-${item.key}`,
+          type: 'cell',
+          index: 1,
+          value: null,
+          level: 0,
+          rendered: null,
+          textValue: item.textValue, // TODO localize?
+          hasChildNodes: false,
+          childNodes: []
+        });
+      }
+      return {
+        type: 'item',
+        childNodes
+      };
+    })
+  }), [collection, isRemovable]);
+
+  let state = useGridState({
+    ...props,
+    collection: gridCollection,
+    focusMode: 'cell'
+  });
+
+  let keyboardDelegate = new GridKeyboardDelegate({
+    collection: state.collection,
+    disabledKeys: state.disabledKeys,
+    ref: domRef,
+    direction,
+    focusMode: 'cell',
+    cycleMode: 'within'
+  });
```

Comment:
Hm strange, I'm able to focus and navigate between the disabled tags in https://reactspectrum.blob.core.windows.net/reactspectrum/fa23d87f3c4910de1fed53dc75b2ef39943be314/storybook/index.html?path=/story/taggroup--disabled. Maybe something with the disabled keys, I would've expected the keyboard delegate to handle that

Reply:
Probably double check your `state.disabledKeys`

Reply:
i was looking at GridKeyboardDelegate and i don't see any logic related to disabled keys. that could be why it's still focusing on the disabled item. will look further